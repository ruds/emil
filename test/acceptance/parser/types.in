foo : 'a;
foo : 'a list;
foo : 'a list list;
foo : bigint;
foo : (string, 'a list) Dict.dict;
foo : ('a list) list;
foo : ('a list);
foo : ((string, ('a list)) Dict.dict);
foo : { bar : (string, 'a list) Dict.dict, baz : bigint, quux : 'a -> string };
foo : bigint * string;
foo : bigint -> string;
foo : bigint * 'a list * (bigint * string) * float;
foo : 'a * 'b -> 'c * bigint * string * (bigint -> float) -> bigint;
foo : { bar : string, bar : string };
