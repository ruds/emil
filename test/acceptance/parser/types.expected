0001 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (VarTypeExpr
                         :id
                         "2761"))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0002 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (TyconTypeExpr
                         :qualifiers
                         ()
                         :identifier
                         "6C697374"
                         :types
                         ((VarTypeExpr
                             :id
                             "2761"))))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0003 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (TyconTypeExpr
                         :qualifiers
                         ()
                         :identifier
                         "6C697374"
                         :types
                         ((TyconTypeExpr
                             :qualifiers
                             ()
                             :identifier
                             "6C697374"
                             :types
                             ((VarTypeExpr
                                 :id
                                 "2761"))))))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0004 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (TyconTypeExpr
                         :qualifiers
                         ()
                         :identifier
                         "626967696E74"
                         :types
                         ()))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0005 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (TyconTypeExpr
                         :qualifiers
                         ("44696374")
                         :identifier
                         "64696374"
                         :types
                         (
                             (TyconTypeExpr
                                 :qualifiers
                                 ()
                                 :identifier
                                 "737472696E67"
                                 :types
                                 ())
                             (TyconTypeExpr
                                 :qualifiers
                                 ()
                                 :identifier
                                 "6C697374"
                                 :types
                                 ((VarTypeExpr
                                     :id
                                     "2761")))
                         )))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0006 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (TyconTypeExpr
                         :qualifiers
                         ()
                         :identifier
                         "6C697374"
                         :types
                         ((TyconTypeExpr
                             :qualifiers
                             ()
                             :identifier
                             "6C697374"
                             :types
                             ((VarTypeExpr
                                 :id
                                 "2761"))))))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0007 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (TyconTypeExpr
                         :qualifiers
                         ()
                         :identifier
                         "6C697374"
                         :types
                         ((VarTypeExpr
                             :id
                             "2761"))))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0008 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (TyconTypeExpr
                         :qualifiers
                         ("44696374")
                         :identifier
                         "64696374"
                         :types
                         (
                             (TyconTypeExpr
                                 :qualifiers
                                 ()
                                 :identifier
                                 "737472696E67"
                                 :types
                                 ())
                             (TyconTypeExpr
                                 :qualifiers
                                 ()
                                 :identifier
                                 "6C697374"
                                 :types
                                 ((VarTypeExpr
                                     :id
                                     "2761")))
                         )))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0009 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (RecordTypeExpr
                         :rows
                         (
                             ("626172"
                               . (TyconTypeExpr
                                     :qualifiers
                                     ("44696374")
                                     :identifier
                                     "64696374"
                                     :types
                                     (
                                         (TyconTypeExpr
                                             :qualifiers
                                             ()
                                             :identifier
                                             "737472696E67"
                                             :types
                                             ())
                                         (TyconTypeExpr
                                             :qualifiers
                                             ()
                                             :identifier
                                             "6C697374"
                                             :types
                                             ((VarTypeExpr
                                                 :id
                                                 "2761")))
                                     )))
                             ("62617A"
                               . (TyconTypeExpr
                                     :qualifiers
                                     ()
                                     :identifier
                                     "626967696E74"
                                     :types
                                     ()))
                             ("71757578"
                               . (FuncTypeExpr
                                     :param
                                     (VarTypeExpr
                                         :id
                                         "2761")
                                     :ret
                                     (TyconTypeExpr
                                         :qualifiers
                                         ()
                                         :identifier
                                         "737472696E67"
                                         :types
                                         ())))
                         )))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0010 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (TupleTypeExpr
                         :types
                         (
                             (TyconTypeExpr
                                 :qualifiers
                                 ()
                                 :identifier
                                 "626967696E74"
                                 :types
                                 ())
                             (TyconTypeExpr
                                 :qualifiers
                                 ()
                                 :identifier
                                 "737472696E67"
                                 :types
                                 ())
                         )))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0011 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (FuncTypeExpr
                         :param
                         (TyconTypeExpr
                             :qualifiers
                             ()
                             :identifier
                             "626967696E74"
                             :types
                             ())
                         :ret
                         (TyconTypeExpr
                             :qualifiers
                             ()
                             :identifier
                             "737472696E67"
                             :types
                             ())))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0012 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (TupleTypeExpr
                         :types
                         (
                             (TyconTypeExpr
                                 :qualifiers
                                 ()
                                 :identifier
                                 "626967696E74"
                                 :types
                                 ())
                             (TyconTypeExpr
                                 :qualifiers
                                 ()
                                 :identifier
                                 "6C697374"
                                 :types
                                 ((VarTypeExpr
                                     :id
                                     "2761")))
                             (TupleTypeExpr
                                 :types
                                 (
                                     (TyconTypeExpr
                                         :qualifiers
                                         ()
                                         :identifier
                                         "626967696E74"
                                         :types
                                         ())
                                     (TyconTypeExpr
                                         :qualifiers
                                         ()
                                         :identifier
                                         "737472696E67"
                                         :types
                                         ())
                                 ))
                             (TyconTypeExpr
                                 :qualifiers
                                 ()
                                 :identifier
                                 "666C6F6174"
                                 :types
                                 ())
                         )))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0013 (DeclTopDecl
         :decl
         (ValDecl
             :bindings
             ((ValBind
                 :pat
                 (IdentifierPattern
                     :qualifiers
                     ()
                     :identifier
                     "6974"
                     :is_op
                     false
                     :is_prefix_op
                     false)
                 :expr
                 (TypedExpr
                     :expr
                     (IdentifierExpr
                         :qualifiers
                         ()
                         :identifier
                         "666F6F"
                         :is_op
                         false
                         :is_prefix_op
                         false)
                     :type
                     (FuncTypeExpr
                         :param
                         (TupleTypeExpr
                             :types
                             (
                                 (VarTypeExpr
                                     :id
                                     "2761")
                                 (VarTypeExpr
                                     :id
                                     "2762")
                             ))
                         :ret
                         (FuncTypeExpr
                             :param
                             (TupleTypeExpr
                                 :types
                                 (
                                     (VarTypeExpr
                                         :id
                                         "2763")
                                     (TyconTypeExpr
                                         :qualifiers
                                         ()
                                         :identifier
                                         "626967696E74"
                                         :types
                                         ())
                                     (TyconTypeExpr
                                         :qualifiers
                                         ()
                                         :identifier
                                         "737472696E67"
                                         :types
                                         ())
                                     (FuncTypeExpr
                                         :param
                                         (TyconTypeExpr
                                             :qualifiers
                                             ()
                                             :identifier
                                             "626967696E74"
                                             :types
                                             ())
                                         :ret
                                         (TyconTypeExpr
                                             :qualifiers
                                             ()
                                             :identifier
                                             "666C6F6174"
                                             :types
                                             ()))
                                 ))
                             :ret
                             (TyconTypeExpr
                                 :qualifiers
                                 ()
                                 :identifier
                                 "626967696E74"
                                 :types
                                 ()))))
                 :rec
                 false))
             :explicit_type_vars
             ()))
0014 ERROR
0015 (EndOfFileTopDecl)
